sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateJoinToEmptyRelation.scala

```diff
@@ -36,22 +36,40 @@ import org.apache.spark.sql.execution.joins.{EmptyHashedRelation, HashedRelation
  */
 object EliminateJoinToEmptyRelation extends Rule[LogicalPlan] {
 
-  private def canEliminate(plan: LogicalPlan, relation: HashedRelation): Boolean = plan match {
-    case LogicalQueryStage(_, stage: BroadcastQueryStageExec) if stage.resultOption.get().isDefined
-      && stage.broadcast.relationFuture.get().value == relation => true
-    case _ => false
+  private def canEliminate(
+      streamedPlan: LogicalPlan,
+      buildPlan: LogicalPlan,
+      relation: HashedRelation): Boolean = {
+    // If streamedSide of the Join contains ShuffleQueryStageExec(canChangeNumPartitions== false)
+    // it can't be rewritten to EmptyRelation because the conversion might lost user specified
+    // number partition information.
+    val immutablePartitionStages = streamedPlan.collect {
+      case lqs @ LogicalQueryStage(_, stage: ShuffleQueryStageExec)
```

Comment:
`LogicalQueryStage.physicalPlan` is not always a query stage. It can be the final agg. It's safer if we do another `collect` here
```
val immutablePartitionStages = streamedPlan.collect {
  case lqs @ LogicalQueryStage(_, physicalPlan: plan) =>
    physicalPlan.collect {
      case s: ShuffleQueryStageExec if !s.shuffle.canChangeNumPartitions => s
    }.nonEmpty
}.filter(_ == true)
```

Reply:
you're right. will update.